File: android/app/src/main/java/tw/stsa/memberapp/feature/events/TicketDocument.kt
package tw.stsa.memberapp.feature.events

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.core.graphics.createBitmap
import com.google.zxing.BarcodeFormat
import com.google.zxing.BinaryBitmap
import com.google.zxing.DecodeHintType
import com.google.zxing.MultiFormatReader
import com.google.zxing.RGBLuminanceSource
import com.google.zxing.common.HybridBinarizer
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import tw.stsa.memberapp.R
import tw.stsa.memberapp.auth.IndicoAuthManager
import tw.stsa.memberapp.feature.checkin.ScannedCode
import tw.stsa.memberapp.net.httpGetBytes
import java.io.File

/**
 * The member's own ticket, read once so the app can show the code itself.
 *
 * Mirrors `ios/MemberApp/Features/Events/TicketDocument.swift`, and for the same
 * reason: Indico hands a member their ticket as a PDF and nothing else. The
 * check-in payload exists as data — `/api/checkin/…` is made of it — but that API
 * is `RHManageEventBase`, an organiser reading a roster, so a member asking for
 * their own gets a 403. The PDF is what they are allowed to have, and the QR is
 * printed on it. So the page is rendered and the code read back off it.
 *
 * [ScannedCode.parse] is the filter, and it is the 報到 scanner's own parser
 * rather than a lookalike: a ticket template can carry other codes, and only one
 * of them is an Indico ticket. When one is found the app draws it itself; when
 * none is, the rendered page is shown as Indico drew it, which still scans.
 *
 * **The one thing iOS does not have to do.** `PDFDocument` there takes bytes;
 * [PdfRenderer] here takes a *seekable file descriptor* and there is no public
 * way around it. So the bytes touch app-private `cacheDir` for the length of one
 * render and are deleted in a `finally` — never external storage, never a
 * MediaStore entry, and gone before the bitmap is handed back. Nothing else is
 * persisted: a ticket QR *is* the credential, which is why [TicketStore] keeps
 * only URLs.
 */
class TicketDocument(context: Context) {

    private val appContext = context.applicationContext

    sealed interface State {
        data object Loading : State
        data class Ready(val ticket: Ticket) : State
        data class Failed(val message: String) : State
    }

    sealed interface Ticket {
        /** The check-in payload, verbatim, for the app to draw. */
        data class Code(val payload: String) : Ticket

        /** The ticket as Indico drew it, for when no code could be read back. */
        data class Page(val image: ImageBitmap) : Ticket
    }

    var state: State by mutableStateOf(State.Loading)
        private set

    suspend fun load(url: String, indico: IndicoAuthManager) {
        state = State.Loading

        state = try {
            val response = httpGetBytes(url, indico.authorizationHeaders())

            // The same test the probe that found this ticket runs, and for the
            // same reason: the client follows redirects, so a request that lost
            // its authorization comes back a perfectly good 200 carrying
            // Indico's login page. Only a PDF body is a ticket.
            val outcome = TicketStore.outcome(response.status, response.contentType)
            if (outcome != TicketStore.Outcome.AVAILABLE) {
                State.Failed(appContext.getString(R.string.ticket_fetch_failed))
            } else {
                read(response.bytes)
                    ?.let { State.Ready(it) }
                    ?: State.Failed(appContext.getString(R.string.ticket_unreadable))
            }
        } catch (error: Exception) {
            State.Failed(error.message ?: appContext.getString(R.string.ticket_fetch_failed))
        }
    }

    /**
     * Rendering a page and searching it for a QR is real work, so it happens off
     * the main thread rather than stalling the screen it arrives behind.
     */
    private suspend fun read(pdf: ByteArray): Ticket? = withContext(Dispatchers.Default) {
        val page = render(pdf) ?: return@withContext null
        code(page)?.let { Ticket.Code(it) } ?: Ticket.Page(page.asImageBitmap())
    }

    private fun render(pdf: ByteArray): Bitmap? {
        // `createTempFile` in cacheDir: app-private, and deleted below whatever
        // happens. See the class note — this is the whole reason it exists.
        val file = File.createTempFile("ticket", ".pdf", appContext.cacheDir)
        return try {
            file.writeBytes(pdf)
            ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY).use { descriptor ->
                PdfRenderer(descriptor).use { renderer ->
                    if (renderer.pageCount == 0) return null
                    renderer.openPage(0).use { page ->
                        // Indico's ticket templates are badge-sized, so the page
                        // is small in points and at its own scale the QR is
                        // barely wider than its own modules — enough to print,
                        // not enough to find again.
                        val bitmap = createBitmap(page.width * 3, page.height * 3)
                        // PdfRenderer draws onto transparency, and a QR read off
                        // a transparent-black field is no QR at all.
                        bitmap.eraseColor(Color.WHITE)
                        page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
                        bitmap
                    }
                }
            }
        } catch (error: Exception) {
            null
        } finally {
            file.delete()
        }
    }

    private fun code(page: Bitmap): String? = runCatching {
        val width = page.width
        val height = page.height
        val pixels = IntArray(width * height)
        page.getPixels(pixels, 0, width, 0, 0, width, height)

        val source = RGBLuminanceSource(width, height, pixels)
        val reader = MultiFormatReader().apply {
            setHints(
                mapOf(
                    DecodeHintType.POSSIBLE_FORMATS to listOf(BarcodeFormat.QR_CODE),
                    // A ticket page is mostly white with one small code on it,
                    // which is exactly the case the fast path gives up on.
                    DecodeHintType.TRY_HARDER to true,
                )
            )
        }

        val text = reader.decodeWithState(BinaryBitmap(HybridBinarizer(source))).text
        // Only an Indico ticket counts. A template may carry a link to the event
        // page or an organiser's own code, and neither is what gets scanned at
        // the door.
        if (ScannedCode.parse(text) is ScannedCode.Ticket) text else null
    }.getOrNull()
}

File: ios/MemberApp/Features/Events/TicketDocument.swift
import CoreImage
import PDFKit
import SwiftUI

/// The member's own ticket, read once so the app can show the code itself.
///
/// Indico hands a member their ticket as a PDF and nothing else. The check-in
/// payload does exist as data — `/api/checkin/…` is made of it — but that API is
/// `RHManageEventBase`, an organiser reading a roster, so a member asking for
/// their own gets a 403. The PDF is what they are allowed to have, and the QR is
/// printed on it. So the page is rendered and the code read back off it.
///
/// `ScannedCode.parse` is the filter, and it is the door scanner's own parser
/// rather than a lookalike: a ticket template can carry other codes — a link to
/// the event page, an organiser's logo — and only one of them is an Indico
/// ticket. When one is found the app draws it itself, large and square-edged, at
/// the size a phone screen is actually scanned from. When none is, the rendered
/// page is shown as Indico drew it, which still scans and still says everything
/// the ticket says.
///
/// Nothing is written to disk and nothing outlives the screen: a ticket QR *is*
/// the credential — whoever holds it can be checked in as that member, which is
/// why `TicketStore` keeps only URLs — so this lives for as long as the view
/// does and no longer.
@Observable
final class TicketDocument {
    enum State {
        case loading
        case ready(Ticket)
        case failed(String)
    }

    enum Ticket {
        /// The check-in payload, verbatim, for the app to draw.
        case code(String)
        /// The ticket as Indico drew it, for when no code could be read back.
        case page(UIImage)
    }

    private(set) var state = State.loading

    func load(from url: URL, using indico: IndicoAuthManager) async {
        state = .loading

        do {
            let request = try indico.authorizedRequest(for: url)
            let (data, response) = try await URLSession.shared.data(for: request)
            let http = response as? HTTPURLResponse

            // The same test the probe that found this ticket runs, and for the
            // same reason: `URLSession` follows redirects, so a request that
            // lost its authorization comes back a perfectly good 200 carrying
            // Indico's login page. Only a PDF body is a ticket.
            let outcome = TicketStore.outcome(
                status: http?.statusCode ?? 0,
                contentType: http?.value(forHTTPHeaderField: "Content-Type")
            )
            guard outcome == .available else { throw LoadError.unavailable }
            guard let ticket = await Self.read(data) else { throw LoadError.unreadable }

            state = .ready(ticket)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    enum LoadError: LocalizedError {
        case unavailable
        case unreadable

        var errorDescription: String? {
            switch self {
            case .unavailable: String(localized: "無法取得票券，稍後再試。")
            case .unreadable: String(localized: "票券格式無法讀取。")
            }
        }
    }

    // MARK: - Reading the page

    /// Rendering a page and searching it for a QR is real work, and it happens
    /// while the screen is pushing in, so it goes off the main actor rather than
    /// stalling the animation it arrives behind.
    @concurrent
    private nonisolated static func read(_ data: Data) async -> Ticket? {
        guard let page = PDFDocument(data: data)?.page(at: 0) else { return nil }

        // Indico's ticket templates are badge-sized, so the page is small in
        // points and at its own scale the QR is barely wider than its own
        // modules — enough to print, not enough to find again. 3× is enough for
        // the detector and cheap enough to do while the screen animates.
        let bounds = page.bounds(for: .mediaBox)
        guard bounds.width > 0, bounds.height > 0 else { return nil }
        let rendered = page.thumbnail(
            of: CGSize(width: bounds.width * 3, height: bounds.height * 3),
            for: .mediaBox
        )

        if let code = code(in: rendered) { return .code(code) }
        return .page(rendered)
    }

    private nonisolated static func code(in image: UIImage) -> String? {
        guard let cgImage = image.cgImage else { return nil }

        let detector = CIDetector(ofType: CIDetectorTypeQRCode,
                                  context: CIContext(),
                                  options: [CIDetectorAccuracy: CIDetectorAccuracyHigh])
        let features = detector?.features(in: CIImage(cgImage: cgImage)) ?? []

        for feature in features {
            guard let message = (feature as? CIQRCodeFeature)?.messageString else { continue }
            if case .ticket = ScannedCode.parse(message) { return message }
        }
        return nil
    }
}
